patroni/ha.py

```diff
@@ -579,6 +626,30 @@ def post_recover(self):
             return 'failed to start postgres'
         return None
 
+    def handle_starting_instance(self):
+        """Starting up PostgreSQL may take a long time. In case we are the leader we may want to
+        fail over to."""
+        # state_handler.state == 'starting' here
+        if self.has_lock():
+            time_left = self.patroni.config['master_start_timeout'] - self.state_handler.time_in_state()
+            if time_left <= 0:
+                if self.is_failover_possible(self.cluster.members):
+                    logger.info("Demoting self because master startup is taking too long")
+                    self.demote_asap()
+                    return 'stopped PostgreSQL because of startup timeout'
+                else:
+                    return 'master start has timed out, but continuing to wait because failover is not possible'
+            else:
+                msg = self.process_manual_failover_from_leader()
```

Comment:
I moved the check for failover key into `process_manual_failover_from_leader` to avoid code duplication. The function can now be called whenever we are ready to handle a manual failover and will return a message if action was taken. I'll add documentation to the function.